%3a117/list-queues-impl.scm

```diff
@@ -0,0 +1,237 @@
+;;;; Implementation of list-queue SRFI
+;;;;
+;;;; Copyright (C) John Cowan (2015, 2016). All Rights Reserved.
+;;;;
+;;;; Permission is hereby granted, free of charge, to any person obtaining a copy of
+;;;; this software and associated documentation files (the "Software"), to deal in
+;;;; the Software without restriction, including without limitation the rights to
+;;;; use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
+;;;; of the Software, and to permit persons to whom the Software is furnished to do
+;;;; so, subject to the following conditions:
+;;;;
+;;;; The above copyright notice and this permission notice shall be included in all
+;;;; copies or substantial portions of the Software.
+;;;;
+;;;; THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+;;;; IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+;;;; FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+;;;; AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+;;;; LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+;;;; OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+;;;; SOFTWARE.
+
+;;;; Later modified from the R7RS SRFI-117 reference implementation for
+;;;; possible inclusion within the R6RS https://github.com/arcfide/chez-srfi.
+
+;;; R7RS shims.  Comment these out on an R7RS system.
+;;; [John Cowan] stole this code from Chibi Scheme, which is BSD-licensed.
+
+(define (make-list n . o)
+  (let ((default (if (pair? o) (car o))))
+    (let lp ((n n) (res '()))
+      (if (<= n 0) res (lp (- n 1) (cons default res))))))
+
+(define (list-copy ls)
+  (let lp ((ls ls) (res '()))
+    (if (pair? ls)
+        (lp (cdr ls) (cons (car ls) res))
+        (append (reverse res) ls))))
+
+(define (list-set! ls k x)
+  (cond ((null? ls) (error "invalid list index"))
+        ((zero? k) (set-car! ls x))
+        (else (list-set! (cdr ls) (- k 1) x))))
+
+;;; This definition is from Chibi's SRFI-1 implementation.
+
+(define (last-pair ls) (if (null? (cdr ls)) ls (last-pair (cdr ls))))
+
+;;; This definition of map! isn't fully SRFI-1 compliant, as it
+;;; handles only unary functions.  You can use SRFI-1's definition
+;;; if you want.
+
+(define (map! f lis)
+  (let lp ((lis lis))
+    (if (pair? lis)
+      (begin
+        (set-car! lis (f (car lis)))
+        (lp (cdr lis))))))
+
+;;; The list-queue record
+;;; The invariant is that either first is (the first pair of) a list
+;;; and last is the last pair, or both of them are the empty list.
+
+(define-record-type
+  (<list-queue> raw-make-list-queue list-queue?)
+  (fields
+    (mutable first get-first set-first!)
+    (mutable last get-last set-last!)))
```

Comment:
Minor change relative to the reference implementation.  Original was https://github.com/scheme-requests-for-implementation/srfi-117/blob/master/list-queues/list-queues-impl.scm#L41.